src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -492,7 +524,103 @@ private final class BitmapIndexedMapNode[K, +V](
+    *
+    * @param key the key to update
+    * @param value the value to set `key` to
+    * @param originalHash key.##
+    * @param keyHash the improved hash
+    * @param shallowlyMutableNodeMap bitmap of child nodes of this node, which can be shallowly mutated
+    *                                during the call to this method
+    *
+    * @return Int which is the bitwise OR of shallowlyMutableNodeMap and any freshly created nodes, which will be
+    *         available for mutations in subsequent calls.
+    */
+  def updateWithShallowMutations[V1 >: V](key: K, value: V1, originalHash: Int, keyHash: Int, shift: Int, shallowlyMutableNodeMap: Int): Int = {
+    val mask = maskFrom(keyHash, shift)
+    val bitpos = bitposFrom(mask)
+
+    if ((dataMap & bitpos) != 0) {
+      val index = indexFrom(dataMap, mask, bitpos)
+      val key0 = getKey(index)
+      val key0UnimprovedHash = getHash(index)
+      if (key0UnimprovedHash == originalHash && key0 == key) {
+        val value0 = this.getValue(index)
+        if (!((key0.asInstanceOf[AnyRef] eq key.asInstanceOf[AnyRef]) && (value0.asInstanceOf[AnyRef] eq value.asInstanceOf[AnyRef]))) {
+          val dataIx = dataIndex(bitpos)
+          val idx = TupleLength * dataIx
+          content(idx + 1) = value
+        }
+        return shallowlyMutableNodeMap
+      } else {
+        val value0 = this.getValue(index)
+        val key0Hash = improve(key0UnimprovedHash)
+
+        val subNodeNew = mergeTwoKeyValPairs(key0, value0, key0UnimprovedHash, key0Hash, key, value, originalHash, keyHash, shift + BitPartitionSize)
+        migrateFromInlineToNodeInPlace(bitpos, key0Hash, subNodeNew)
+        return shallowlyMutableNodeMap | bitpos
+      }
+    }
+
+    if ((nodeMap & bitpos) != 0) {
+      val index = indexFrom(nodeMap, mask, bitpos)
+      val subNode = this.getNode(index)
+      val subNodeSize = subNode.size
+      val subNodeCachedJavaKeySetHashCode = subNode.cachedJavaKeySetHashCode
+
+      var returnShallowlyMutableNodeMap = shallowlyMutableNodeMap
+
+      val subNodeNew: MapNode[K, V1] = subNode match {
+        case subNodeBm: BitmapIndexedMapNode[K, V] if (bitpos & shallowlyMutableNodeMap) != 0 =>
+          subNodeBm.updateWithShallowMutations(key, value, originalHash, keyHash, shift + BitPartitionSize, 0)
+          subNodeBm
+        case _ =>
+          val result = subNode.updated(key, value, originalHash, keyHash, shift + BitPartitionSize)
+          if (result ne subNode) {
+            returnShallowlyMutableNodeMap |= bitpos
+          }
+          result
+      }
+
+      this.content(this.content.length - 1 - this.nodeIndex(bitpos)) = subNodeNew
+      this.size = this.size - subNodeSize + subNodeNew.size
+      this.cachedJavaKeySetHashCode = this.cachedJavaKeySetHashCode - subNodeCachedJavaKeySetHashCode + subNodeNew.cachedJavaKeySetHashCode
+      return returnShallowlyMutableNodeMap
+    }
+
+    {
```

Comment:
This can be an `else` since you can't get here via the other branches; all paths return.

Reply:
Shortened